submission/persistence-filesystem/src/main/kotlin/ac/uk/ebi/biostd/persistence/filesystem/fire/FireService.kt

```diff
@@ -0,0 +1,80 @@
+package ac.uk.ebi.biostd.persistence.filesystem.fire
+
+import ebi.ac.uk.extended.model.ExtFile
+import ebi.ac.uk.extended.model.ExtFileType
+import ebi.ac.uk.extended.model.ExtSubmission
+import ebi.ac.uk.extended.model.FireFile
+import ebi.ac.uk.extended.model.NfsFile
+import ebi.ac.uk.io.ext.md5
+import org.zeroturnaround.zip.ZipUtil
+import uk.ac.ebi.fire.client.integration.web.FireWebClient
+import uk.ac.ebi.fire.client.model.FireApiFile
+import java.io.File
+import java.nio.file.Files
+
+class FireService(
+    private val fireWebClient: FireWebClient,
+    private val fireTempDirPath: File,
+) {
+
+    fun getOrPersist(sub: ExtSubmission, file: ExtFile): FireFile {
+        return when (file) {
+            is FireFile -> reuseFireFile(sub, file)
+            is NfsFile -> getOrPersist(sub, file)
+        }
+    }
+
+    fun cleanFtp(sub: ExtSubmission) {
+        fireWebClient
+            .findByAccNo(sub.accNo)
+            .forEach { fireWebClient.unsetPath(it.fireOid) }
+    }
+
+    private fun reuseFireFile(sub: ExtSubmission, fireFile: FireFile): FireFile {
+        val newPath = "${sub.relPath}/${fireFile.relPath}"
+        fireWebClient.setPath(fireFile.fireId, newPath)
+        return fireFile
+    }
+
+    private fun getOrPersist(sub: ExtSubmission, nfsFile: NfsFile): FireFile {
+        return when (val record = fireWebClient.findByPath("${sub.relPath}/${nfsFile.relPath}")) {
+            null -> if (nfsFile.file.isDirectory) saveDirectory(sub, nfsFile) else saveFile(sub, nfsFile)
+            else -> asFireFile(nfsFile, record, nfsFile.type)
+        }
+    }
+
+    private fun saveFile(sub: ExtSubmission, nfsFile: NfsFile): FireFile {
+        val fireFile = fireWebClient.save(nfsFile.file, nfsFile.md5)
+        fireWebClient.setBioMetadata(fireFile.fireOid, sub.accNo, nfsFile.type.value, published = false)
+        fireWebClient.setPath(fireFile.fireOid, "${sub.relPath}/${nfsFile.relPath}")
+        return asFireFile(nfsFile, fireFile, ExtFileType.FILE)
+    }
+
+    private fun saveDirectory(sub: ExtSubmission, nfsFile: NfsFile): FireFile {
+        val directory = compress(sub, nfsFile.file)
+        val fireFile = fireWebClient.save(directory, directory.md5())
+        fireWebClient.setBioMetadata(fireFile.fireOid, sub.accNo, nfsFile.type.value, published = false)
+        fireWebClient.setPath(fireFile.fireOid, "${sub.relPath}/${nfsFile.relPath}.zip")
+        return asFireFile(nfsFile, fireFile, ExtFileType.DIR)
+    }
+
+    private fun compress(sub: ExtSubmission, file: File): File {
+        val tempFolder = fireTempDirPath.resolve("${sub.accNo}/${sub.version}")
+        tempFolder.mkdirs()
+
+        val target = tempFolder.resolve(file.name)
+        Files.deleteIfExists(target.toPath())
```

Comment:
In the previous implementation, the directory was only compressed once. We did it like this so we would avoid compressing every time in case the persistence process fails since it should already be in the temp folder.

Do we really wanna change this?

Reply:
yes it is intencional, the reasoing is that there is not garantee that previos directory zip is the same as new one (as folder content may have change) so it is safer to re compress it.